### ISSUE TITLE
IllegalStateException: ... already opened with different schema version.

### ISSUE THREAD

**Issue by moldovanpeter89** [realm/realm-kotlin#1293]
Hello,

I have an issue opening Realm-Kotlin database after an automatic migration on iOS application which uses also RealmSwift.
I can't provide the logs tested on the live app.

Context:
- We have an android app using Realm-Java.
- We have an iOS app using RealmSwift.
- We want to migrate both apps to Realm-Kotlin, but it's not possible to do it in one shot.

We have a situation where the Property for the same Schema is a bit different between platforms, this example simulates our case:

**Android**
```
class Sample : RealmObject {

    @PrimaryKey
    var id: String? = null

    var name: String? = null

    var street: String? = null
}
```

**iOS**
```
final class Sample: Object {

    @objc dynamic var id: String = ""
    override public static func primaryKey() -> String? {
        return "id"
    }

    @objc dynamic var name: String?

    @objc dynamic var address: String?
}
```
We can see the difference is street and address in the Sample Schema.

What I chose to do is "manually" merge the Schema for the shared KMM Schema to be "compatible" with both platforms:

**KMM**
```
class Sample : RealmObject {

    @PrimaryKey
    var id: String? = null

    var name: String? = null

    var street: String? = null

    var address: String? = null
}
```
after this, I Incremented the Schema Version when opening the Realm-Kotlin / platform.

- On Android, the automatic migration is triggered and the "Shared" database can be opened and queried.
- On iOS after the migration we have IllegalStateException: .... already opened with different schema version. (Trimmed the log).

I checked and according to the official documentation of RealmSwift, the database is kept open until ARC handles it. 
I am unsure if this is the correct approach and if there is a way to handle this correctly. I could not find information about it.

The realm versions that were used is to have the same RealmCore version (different issue on iOS):

- realm-swift-sdk: 10.35.0
- realm-kotlin-sdk: io.realm.kotlin:library-base:1.7.0-SNAPSHOT

**Comment by clementetb** [comment]
It seems that RealmSwift has opened the realm file before Realm-Kotlin, and both SDKs use a configuration with a different schema version.

Aligning the version number should allow opening the Realm. However, we don't [recommend](https://github.com/realm/realm-kotlin/issues/1278#issuecomment-1425866709) opening the same file with two different SDKs. Depending on the SDK version they might be using a different core version that might lead to corruption, and we don't guarantee that notifications on data change as it has not been tested.

**Comment by moldovanpeter89** [comment]
@clementetb, 
yes, you are right that there is a difference in version number, I tried to explain that because of inconsistency between the Schemas, I am migrating to a "common" Schema that can be used by both platforms.
On Android side, this is working because we close the "file" after the transactions, but on the iOS part is not because the previous one remains open.

I am sure that we don't live in an "ideal" world and this case should not be unique 😿 , so for the existing applications which want to migrate to KMM with Realm-Kotlin and used the "native" SDKs of realm before... I don't see a good solution at the moment on how to do the "extraction".
The workaround that I tried is using expect/actual and it works, but it brings a big overhead.

>However, we don't https://github.com/realm/realm-kotlin/issues/1278#issuecomment-1425866709 opening the same file with two different SDKs.

I did understand it before but unfortunately, we can't avoid it 🥲 .


Thank you for the quick reply.
